# Big Data Activity - cars.csv

This notebook explores, cleans and transforms **cars.csv** (428 cars, 15 columns) using pandas and numpy. Each section has an explanation, the code, its output, and a short read of what the output means.

## a. Load the data and preview it

I load the CSV with `pd.read_csv()` and look at the top rows, bottom rows and a random sample to get a feel for the table.

In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv('cars.csv')
df.head()

,Make,Model,Type,Origin,DriveTrain,MSRP,Invoice,EngineSize,Cylinders,Horsepower,MPG_City,MPG_Highway,Weight,Wheelbase,Length
0,Acura,MDX,SUV,Asia,All,36945,33337,3.5,6.0,265.0,17.0,23.0,4451.0,106.0,189.0
1,Acura,RSX Type S 2dr,Sedan,Asia,Front,23820,21761,2.0,4.0,200.0,24.0,31.0,2778.0,101.0,172.0
2,Acura,TSX 4dr,Sedan,Asia,Front,26990,24647,2.4,4.0,200.0,22.0,29.0,3230.0,105.0,183.0
3,Acura,TL 4dr,Sedan,Asia,Front,33195,30299,3.2,6.0,270.0,20.0,28.0,NaN,108.0,186.0
4,Acura,3.5 RL 4dr,Sedan,Asia,Front,43755,39014,3.5,6.0,225.0,18.0,24.0,3880.0,115.0,197.0


In [2]:
df.tail()

,Make,Model,Type,Origin,DriveTrain,MSRP,Invoice,EngineSize,Cylinders,Horsepower,MPG_City,MPG_Highway,Weight,Wheelbase,Length
423,Volvo,C70 LPT convertible 2dr,Sedan,Europe,Front,40565,38203,2.4,5.0,197.0,21.0,28.0,3450.0,105.0,186.0
424,Volvo,C70 HPT convertible 2dr,Sedan,Europe,Front,42565,40083,2.3,5.0,242.0,20.0,26.0,3450.0,105.0,186.0
425,Volvo,S80 T6 4dr,Sedan,Europe,Front,45210,42573,2.9,6.0,268.0,19.0,26.0,3653.0,110.0,190.0
426,Volvo,V40,Wagon,Europe,Front,26135,24641,1.9,4.0,170.0,22.0,29.0,2822.0,101.0,180.0
427,Volvo,XC70,Wagon,Europe,All,35145,33112,2.5,5.0,208.0,20.0,27.0,3823.0,109.0,186.0


In [3]:
df.sample(5, random_state=1)

,Make,Model,Type,Origin,DriveTrain,MSRP,Invoice,EngineSize,Cylinders,Horsepower,MPG_City,MPG_Highway,Weight,Wheelbase,Length
227,Lexus,SC 430 convertible 2dr,Sports,Asia,Rear,63200,55063,4.3,8.0,300.0,18.0,23.0,3840.0,103.0,178.0
208,Kia,Spectra 4dr,Sedan,Asia,Front,12360,11630,1.8,4.0,124.0,24.0,32.0,2661.0,101.0,178.0
164,Honda,Odyssey EX,Sedan,Asia,Front,27450,24744,3.5,6.0,240.0,18.0,25.0,4365.0,118.0,201.0
277,Mercury,Mountaineer,SUV,USA,Front,29995,27317,4.0,6.0,210.0,16.0,21.0,4374.0,114.0,190.0
62,Cadillac,Escalade EXT,Truck,USA,All,52975,48541,6.0,8.0,345.0,13.0,17.0,5879.0,130.0,221.0


**Reading it:** The file loads into 428 rows. `head()` starts at Acura and `tail()` ends at Volvo, so the data is sorted alphabetically by Make. The random sample confirms each row is a single car with pricing (MSRP, Invoice) and specs (engine, horsepower, mpg, weight, size).

## b. Structure of the dataset

Here I check the shape, column names, dtypes, a full `info()`, and `describe()` for the numeric summary.

In [4]:
print('Shape:', df.shape)
print('Columns:', list(df.columns))

Shape: (428, 15)
Columns: ['Make', 'Model', 'Type', 'Origin', 'DriveTrain', 'MSRP', 'Invoice', 'EngineSize', 'Cylinders', 'Horsepower', 'MPG_City', 'MPG_Highway', 'Weight', 'Wheelbase', 'Length']


In [5]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 428 entries, 0 to 427
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Make         428 non-null    str    
 1   Model        428 non-null    str    
 2   Type         428 non-null    str    
 3   Origin       428 non-null    str    
 4   DriveTrain   428 non-null    str    
 5   MSRP         428 non-null    int64  
 6   Invoice      428 non-null    int64  
 7   EngineSize   425 non-null    float64
 8   Cylinders    425 non-null    float64
 9   Horsepower   423 non-null    float64
 10  MPG_City     426 non-null    float64
 11  MPG_Highway  426 non-null    float64
 12  Weight       423 non-null    float64
 13  Wheelbase    426 non-null    float64
 14  Length       427 non-null    float64
dtypes: float64(8), int64(2), str(5)
memory usage: 50.3 KB


In [6]:
df.dtypes

Make               str
Model              str
Type               str
Origin             str
DriveTrain         str
MSRP             int64
Invoice          int64
EngineSize     float64
Cylinders      float64
Horsepower     float64
MPG_City       float64
MPG_Highway    float64
Weight         float64
Wheelbase      float64
Length         float64
dtype: object

In [7]:
df.describe()

,MSRP,Invoice,EngineSize,Cylinders,Horsepower,MPG_City,MPG_Highway,Weight,Wheelbase,Length
count,428.000000,428.000000,425.000000,425.000000,423.000000,426.000000,426.000000,423.000000,426.000000,427.000000
mean,32774.855140,30014.700935,3.195294,5.811765,215.576832,20.065728,26.840376,3578.229314,108.166667,186.381733
std,19431.716674,17642.117750,1.111932,1.557802,71.978098,5.249574,5.749505,759.034549,8.327582,14.369110
min,10280.000000,9875.000000,1.300000,3.000000,73.000000,10.000000,12.000000,1850.000000,89.000000,143.000000
25%,20334.250000,18866.000000,2.300000,4.000000,165.000000,17.000000,24.000000,3103.000000,103.000000,178.000000
50%,27635.000000,25294.500000,3.000000,6.000000,210.000000,19.000000,26.000000,3476.000000,107.000000,187.000000
75%,39205.000000,35710.250000,3.900000,6.000000,252.500000,21.750000,29.000000,3978.500000,112.000000,194.000000
max,192465.000000,173560.000000,8.300000,12.000000,500.000000,60.000000,66.000000,7190.000000,144.000000,238.000000


**Reading it:** 428 rows x 15 columns. Five columns are text (Make, Model, Type, Origin, DriveTrain) and ten are numeric. MSRP and Invoice are integers; the other numeric columns are floats because they contain missing values (a NaN forces the column to float). In `describe()` the count row is below 428 for several columns, which is the first sign of missing data, and MSRP runs from about 10,280 to 192,465.

## c. Missing values

I count missing values per column with `isna().sum()`, then fill the numeric columns with their **median** using `fillna()`.

In [8]:
df.isna().sum()

Make           0
Model          0
Type           0
Origin         0
DriveTrain     0
MSRP           0
Invoice        0
EngineSize     3
Cylinders      3
Horsepower     5
MPG_City       2
MPG_Highway    2
Weight         5
Wheelbase      2
Length         1
dtype: int64

I fill with the **median** rather than the mean because the numeric columns are skewed - a few luxury and sports cars have very high price, horsepower and weight, which drags the mean upward. The median sits in the middle of the data and is not distorted by those outliers, so it is a safer stand-in value.

In [9]:
num_cols = df.select_dtypes(include='number').columns
df[num_cols] = df[num_cols].fillna(df[num_cols].median())
df.isna().sum()

Make           0
Model          0
Type           0
Origin         0
DriveTrain     0
MSRP           0
Invoice        0
EngineSize     0
Cylinders      0
Horsepower     0
MPG_City       0
MPG_Highway    0
Weight         0
Wheelbase      0
Length         0
dtype: int64

**Reading it:** Eight numeric columns had gaps (Horsepower and Weight 5 each, EngineSize and Cylinders 3 each, MPG_City, MPG_Highway and Wheelbase 2 each, Length 1) - 23 missing cells in total. After the median fill, `isna().sum()` is 0 everywhere, so the dataset is complete.

## d. Duplicate rows

I check for duplicate rows with `duplicated()` and remove any with `drop_duplicates()`.

In [10]:
print('Duplicate rows:', df.duplicated().sum())
before = df.shape[0]
df = df.drop_duplicates()
print('Rows before:', before, '-> Rows after:', df.shape[0])

Duplicate rows: 0
Rows before: 428 -> Rows after: 428


**Reading it:** `duplicated().sum()` is 0, so there are no repeated rows. `drop_duplicates()` removes nothing and the row count stays 428 - I run it anyway to show the check was done.

## e. Categorical columns and grouping

I use `value_counts()` on Type and Origin, then `groupby()` to compare average, min and max MSRP and Horsepower by Origin and by Type.

In [11]:
df['Type'].value_counts()

Type
Sedan     262
SUV        60
Sports     49
Wagon      30
Truck      24
Hybrid      3
Name: count, dtype: int64

In [12]:
df['Origin'].value_counts()

Origin
Asia      158
USA       147
Europe    123
Name: count, dtype: int64

In [13]:
df.groupby('Origin')[['MSRP','Horsepower']].agg(['mean','min','max']).round(1)

MSRP                Horsepower              
           mean    min     max       mean    min    max
Origin                                                 
Asia    24741.3  10280   89765      190.3   73.0  340.0
Europe  48349.8  16999  192465      252.1  100.0  493.0
USA     28377.4  10995   81795      212.0  103.0  500.0

In [14]:
df.groupby('Type')[['MSRP','Horsepower']].agg(['mean','min','max']).round(1)

MSRP                Horsepower              
           mean    min     max       mean    min    max
Type                                                   
Hybrid  19920.0  19110   20510       92.0   73.0  110.0
SUV     34790.2  17163   76870      234.0  130.0  340.0
Sedan   29773.6  10280  128420      201.8  100.0  493.0
Sports  53387.1  18345  192465      284.1  138.0  500.0
Truck   24941.4  12800   52975      224.8  142.0  345.0
Wagon   28840.5  11905   60670      191.7  104.0  340.0

**Reading it:** Most cars are Sedans (262), then SUV (60), Sports (49), Wagon (30), Truck (24) and only 3 Hybrids. By origin the split is fairly even: Asia 158, USA 147, Europe 123. On price, European cars are the most expensive on average (about 48,350 MSRP) versus USA (about 28,377) and Asia (about 24,741). By type, Sports cars carry the highest average horsepower.

## f. Selecting columns and filtering rows

I select a few columns, then filter rows two ways: SUVs, and cars with Horsepower above 250.

In [15]:
df[['Make','Model','Type','MSRP','Horsepower']].head()

,Make,Model,Type,MSRP,Horsepower
0,Acura,MDX,SUV,36945,265.0
1,Acura,RSX Type S 2dr,Sedan,23820,200.0
2,Acura,TSX 4dr,Sedan,26990,200.0
3,Acura,TL 4dr,Sedan,33195,270.0
4,Acura,3.5 RL 4dr,Sedan,43755,225.0


In [16]:
suvs = df[df['Type'] == 'SUV']
print('Number of SUVs:', suvs.shape[0])
suvs[['Make','Model','MSRP','Horsepower']].head()

Number of SUVs: 60


,Make,Model,MSRP,Horsepower
0,Acura,MDX,36945,265.0
26,BMW,X3 3.0i,37000,225.0
27,BMW,X5 4.4i,52195,325.0
46,Buick,Rainier,37895,275.0
47,Buick,Rendezvous CX,26545,185.0


In [17]:
powerful = df[df['Horsepower'] > 250]
print('Cars with Horsepower > 250:', powerful.shape[0])
powerful[['Make','Model','Horsepower','MSRP']].head()

Cars with Horsepower > 250: 106


,Make,Model,Horsepower,MSRP
0,Acura,MDX,265.0,36945
3,Acura,TL 4dr,270.0,33195
6,Acura,NSX coupe 2dr manual S,290.0,89765
17,Audi,A6 4.2 Quattro 4dr,300.0,49690
18,Audi,A8 L Quattro 4dr,330.0,69190


**Reading it:** Column selection keeps just the fields I care about. Filtering on Type == 'SUV' returns the 60 SUVs, and filtering on Horsepower > 250 returns the higher-performance cars (count shown above) - mostly the pricier sports and luxury models.

## g. New calculated variables

I create `Markup` (MSRP minus Invoice) and `Power_to_Weight` (Horsepower divided by Weight).

In [18]:
df['Markup'] = df['MSRP'] - df['Invoice']
df['Power_to_Weight'] = df['Horsepower'] / df['Weight']
df[['Make','Model','MSRP','Invoice','Markup','Horsepower','Weight','Power_to_Weight']].head()

,Make,Model,MSRP,Invoice,Markup,Horsepower,Weight,Power_to_Weight
0,Acura,MDX,36945,33337,3608,265.0,4451.0,0.059537
1,Acura,RSX Type S 2dr,23820,21761,2059,200.0,2778.0,0.071994
2,Acura,TSX 4dr,26990,24647,2343,200.0,3230.0,0.061920
3,Acura,TL 4dr,33195,30299,2896,270.0,3476.0,0.077675
4,Acura,3.5 RL 4dr,43755,39014,4741,225.0,3880.0,0.057990


**Reading it:** `Markup` is the gap between sticker price and dealer invoice - roughly the dealer margin, and it grows with price. `Power_to_Weight` is horsepower per pound; a higher value means a car is quick for its weight, so light sports cars score highest and heavy SUVs lowest.

## h. Unit conversions

I convert MSRP from USD to Mexican pesos, MPG_City to litres/100km, and Weight from pounds to kilograms.

In [19]:
usd_to_mxn = 18.5  # assumed exchange rate
df['MSRP_MXN'] = df['MSRP'] * usd_to_mxn
df['L_per_100km_City'] = 235.215 / df['MPG_City']
df['Weight_kg'] = df['Weight'] * 0.453592
df[['Model','MSRP','MSRP_MXN','MPG_City','L_per_100km_City','Weight','Weight_kg']].head()

,Model,MSRP,MSRP_MXN,MPG_City,L_per_100km_City,Weight,Weight_kg
0,MDX,36945,683482.5,17.0,13.836176,4451.0,2018.937992
1,RSX Type S 2dr,23820,440670.0,24.0,9.800625,2778.0,1260.078576
2,TSX 4dr,26990,499315.0,22.0,10.691591,3230.0,1465.102160
3,TL 4dr,33195,614107.5,20.0,11.760750,3476.0,1576.685792
4,3.5 RL 4dr,43755,809467.5,18.0,13.067500,3880.0,1759.936960


**Reading it:** Prices are now in pesos using an assumed rate of 18.5 MXN per USD (easy to update if the rate moves). L/100km is the inverse of MPG, so here a lower number means better fuel economy, and weight is in kilograms for the metric system.

## i. Conditional + for loop: price bands

I write a conditional that labels a car Economy, Mid or Premium by MSRP, then loop over every row to tag the whole dataset.

In [20]:
def price_band(msrp):
    if msrp < 20000:
        return 'Economy'
    elif msrp <= 40000:
        return 'Mid'
    else:
        return 'Premium'

labels = []
for _, row in df.iterrows():
    labels.append(price_band(row['MSRP']))

df['Price_Band'] = labels
df['Price_Band'].value_counts()

Price_Band
Mid        228
Premium    102
Economy     98
Name: count, dtype: int64

In [21]:
df[['Make','Model','MSRP','Price_Band']].head(10)

,Make,Model,MSRP,Price_Band
0,Acura,MDX,36945,Mid
1,Acura,RSX Type S 2dr,23820,Mid
2,Acura,TSX 4dr,26990,Mid
3,Acura,TL 4dr,33195,Mid
4,Acura,3.5 RL 4dr,43755,Premium
5,Acura,3.5 RL w/Navigation 4dr,46100,Premium
6,Acura,NSX coupe 2dr manual S,89765,Premium
7,Audi,A4 1.8T 4dr,25940,Mid
8,Audi,A41.8T convertible 2dr,35940,Mid
9,Audi,A4 3.0 4dr,31840,Mid


**Reading it:** The `for` loop runs the conditional on each of the 428 cars and stores the label. The counts show how the market splits into Economy (under 20,000), Mid (20,000-40,000) and Premium (over 40,000), with most cars in the economy and mid bands and fewer in Premium.

## Summary

I loaded cars.csv, checked its structure, filled 23 missing numeric values with medians, confirmed there were no duplicates, explored the categories and groups, filtered and selected rows, built new variables (Markup, Power_to_Weight), converted units for a Mexican context, and used a loop with a conditional to band every car by price.